### ISSUE TITLE
SqlDataReader.GetFieldValue{,Async}() should work for Stream and TextReader

### ISSUE THREAD

**Issue by roji** [dotnet/SqlClient#27]
Following #30958, where @Wraith2 added the ability to retrieve an XmlReader via `SqlDataReader.GetFieldValue()` and its async counterpart, we should do the same for Stream and TextReader. These can already be retrieved via `GetStream()` and `GetTextReader()` respectively.

The main motivation behind this would be to allow getting Reader/TextReader asynchronously without introducing new `GetReaderAsync()` and `GetTextReaderAsync()` methods. See https://github.com/dotnet/corefx/issues/35012#issuecomment-484440830 and below for previous discussion.

**Comment by Wraith2** [comment]
@roji i've got this working on netcore but because of the different netcore and netfx codebases it's more difficult to get it working on netfx. Since this would primarily be useful from EF and as far as a know modern EF only works on core or net5 onwards is there a need to implement on netfx immediately? It will eventually be implemented through the merging of the codebases.

**Comment by roji** [comment]
(sorry for the slow response, back from vacation)

I don't see any particular problem with deferring this for netfx (apart from the general undesirability of having a behavior discrepancy between the versions of the driver); this just adds a new capability which didn't exist before, and it's unlikely anyone's going to be porting new code using it from netcore to netfx...

**Comment by Wraith2** [comment]
Looking at how it's all structured a related question occurred to me. What should happen for `GetFieldValue<int?>` if it were supported? Would we map language null to db null in the case and allow users to avoid the IsDbNull check? Would that be useful or confusing to users? Do other providers support that sort of thing?

**Comment by roji** [comment]
Yeah, `GetFieldValue<int?>` is something we actually implemented in Npgsql to make things slightly easier - even though it doesn't fully respect the DBNull/null distinction. Unfortunately, the way nullable reference types work, it would be impossible for an ADO.NET provider to distinguish between `GetFieldValue<string>` (which should throw on database null) and `GetFieldValue<string?>` which shouldn't.

**Comment by Wraith2** [comment]
streams strings and any other reference type such as UDT's wouldn't work out at all. As you say. 
The standard value types could have nullable versions added. If npgsql does it then it might be worth opening an issue for SqlClient to do the same and possibly ask if mysql and sqlite are interested in the same extension.
